### PR TITLE
collect cpu/mem usage and podStatistics for metrics server

### PR DESCRIPTION
**Pull request by GeorgeGuo2018**
#### What type of PR is this?
kind/feature
#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: deploy/agent/clusternet_agent_deployment.yaml [hunk 1]
```diff
@@ -63,5 +63,6 @@ spec:
             - --cluster-reg-token=$(REG_TOKEN)
             - --cluster-reg-parent-url=$(PARENT_URL)
             - --cluster-sync-mode=Dual
+            - --enable-metric-server=false
```
Comment (dixudx) [inline deploy/agent/clusternet_agent_deployment.yaml:66]: Please submit a separate PR on this change.

This flag is not available in current images.

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -105,6 +106,17 @@ func NewAgent(registrationOpts *ClusterRegistrationOptions, controllerOpts *util
 	}
 	// create clientset for child cluster
 	childKubeClientSet := kubernetes.NewForConfigOrDie(childKubeConfig)
+
+	// create metrics client for child cluster
+	var metricClient *metricsv.Clientset
```
Comment (dixudx) [inline pkg/agent/agent.go:111]:
```go
metricClient := metricsv.NewForConfigOrDie(childKubeConfig)
```

File: pkg/agent/constant.go [hunk 3]
```diff
@@ -49,6 +49,9 @@ const (
 
 	// ClusterStatusCollectFrequency flag specifies the cluster status collecting frequency
 	ClusterStatusCollectFrequency = "cluster-status-collect-frequency"
+
+	// EnableMetricServer flag specifies whether collect metrics from metricServer for the cluster
+	EnableMetricServer = "enable-metric-server"
```
Comment (dixudx) [inline pkg/agent/constant.go:54]: s/EnableMetricServer/UseMetricsServer

File: pkg/agent/constant.go [hunk 4]
```diff
@@ -49,6 +49,9 @@ const (
 
 	// ClusterStatusCollectFrequency flag specifies the cluster status collecting frequency
 	ClusterStatusCollectFrequency = "cluster-status-collect-frequency"
+
+	// EnableMetricServer flag specifies whether collect metrics from metricServer for the cluster
```
Comment (dixudx) [inline pkg/agent/constant.go:53]: whether to collect metrics from metrics server running in the cluster

File: pkg/agent/option.go [hunk 5]
```diff
@@ -59,6 +59,9 @@ type ClusterRegistrationOptions struct {
 	// PredictorAddress specifies the address of predictor
 	PredictorAddress string
 
+	// donot collect metrics from metric-server by default
+	MetricServer bool
```
Comment (dixudx) [inline pkg/agent/option.go:63]: s/MetricServer/UseMetricsServer

File: pkg/agent/constant.go (hunk 3 above)
Comment (dixudx) [reply pkg/agent/constant.go:54]: s/enable-metric-server/use-metrics-server

File: pkg/agent/option.go [hunk 6]
```diff
@@ -59,6 +59,9 @@ type ClusterRegistrationOptions struct {
 	// PredictorAddress specifies the address of predictor
 	PredictorAddress string
 
+	// donot collect metrics from metric-server by default
```
Comment (dixudx) [inline pkg/agent/option.go:62]: UseMetricsServer specifies whether to collect metrics from metrics server

File: pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go [hunk 7]
```diff
@@ -51,6 +53,7 @@ type Controller struct {
 	appPusherEnabled   bool
 	useSocket          bool
 	predictorEnable    bool
+	metricServerEnable bool
```
Comment (dixudx) [inline pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go:56]: s/metricServerEnable/metricsServerEnabled

File: pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go [hunk 8]
```diff
@@ -216,6 +232,50 @@ func getNodeStatistics(nodes []*corev1.Node) (nodeStatistics clusterapi.NodeStat
...
 
+// getPodStatistics returns the PodStatistics in the cluster
+// get pods num in running conditions and the total pods num in the cluster
+func getPodStatistics(clientset *metricsv.Clientset) (podStatistics clusterapi.PodStatistics) {
+	if clientset == nil {
+		klog.Warningf("empty metris client, will return directly ")
+		return
+	}
+	podStatistics = clusterapi.PodStatistics{}
+	podMetricsList, err := clientset.MetricsV1beta1().PodMetricses(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
```
Comment (dixudx) [inline pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go:243]: We need to bind a rule on `podMetrics` for `clusternet-agent`.

File: pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go [hunk 9]
```diff
@@ -216,6 +232,50 @@ func getNodeStatistics(nodes []*corev1.Node) (nodeStatistics clusterapi.NodeStat
...
+	if err != nil {
+		klog.Warningf("failed to list podMetris with err: %v", err.Error)
+		return
+	}
+	for _, item := range podMetricsList.Items {
+		if len(item.Containers) != 0 {
+			podStatistics.RunningPods += 1
```
Comment (dixudx) [inline pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go:250]: This does not seem right.

File: pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go [hunk 10]
```diff
@@ -216,6 +232,50 @@ func getNodeStatistics(nodes []*corev1.Node) (nodeStatistics clusterapi.NodeStat
...
+
+// getResourceUsage returns the ResourceUsage in the cluster
+// get cpu(m) and memory(Mi) used
+func getResourceUsage(clientset *metricsv.Clientset) (resourceUsage clusterapi.ResourceUsage) {
+	if clientset == nil {
+		klog.Warningf("empty metris client, will return directly ")
+		return
+	}
+	resourceUsage = clusterapi.ResourceUsage{}
+	nodeMetricsList, err := clientset.MetricsV1beta1().NodeMetricses().List(context.TODO(), metav1.ListOptions{})
```
Comment (dixudx) [inline pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go:266]: We need to bind a rule on `nodeMetrics` for `clusternet-agent`.

**Review by dixudx — CHANGES_REQUESTED** [review]
We're getting close.

Thanks for your contribution.

File: deploy/agent/clusternet_agent_deployment.yaml (hunk 1 above)
Comment (GeorgeGuo2018) [reply deploy/agent/clusternet_agent_deployment.yaml:66]: ok

File: pkg/agent/agent.go (hunk 2 above)
Comment (GeorgeGuo2018) [reply pkg/agent/agent.go:111]: ok

File: pkg/agent/constant.go (hunk 3 above)
Comment (GeorgeGuo2018) [reply pkg/agent/constant.go:54]: ok

File: pkg/agent/constant.go (hunk 4 above)
Comment (GeorgeGuo2018) [reply pkg/agent/constant.go:53]: ok

File: pkg/agent/option.go (hunk 5 above)
Comment (GeorgeGuo2018) [reply pkg/agent/option.go:63]: ok

File: pkg/agent/option.go (hunk 6 above)
Comment (GeorgeGuo2018) [reply pkg/agent/option.go:62]: ok

File: pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go (hunk 7 above)
Comment (GeorgeGuo2018) [reply pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go:56]: ok

File: pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go [hunk 11]
```diff
@@ -38,11 +38,13 @@ import (
 	clusterapi "github.com/clusternet/clusternet/pkg/apis/clusters/v1beta1"
 	"github.com/clusternet/clusternet/pkg/features"
 	"github.com/clusternet/clusternet/pkg/known"
+	metricsv "k8s.io/metrics/pkg/client/clientset/versioned"
```
Comment (dixudx) [inline pkg/controllers/clusters/clusterstatus/clusterstatus_controller.go:41]: move to L36, and re-format

**Review by dixudx — APPROVED** [review]
